downscale/02_extract_sipnet_output.R

```diff
@@ -0,0 +1,301 @@
+#' | TotLivBiom                    | Total living biomass                     |
+#' | TotSoilCarb                   | Total Soil Carbon                        |
+#' | Qle                           | Latent heat                              |
+#' | Transp                        | Total transpiration                      |
+#' | SoilMoist                     | Average Layer Soil Moisture              |
+#' | SoilMoistFrac                 | Average Layer Fraction of Saturation     |
+#' | SWE                           | Snow Water Equivalent                    |
+#' | litter_carbon_content         | Litter Carbon Content                    |
+#' | litter_mass_content_of_water  | Average layer litter moisture            |
+#' | LAI                           | Leaf Area Index                          |
+#' | fine_root_carbon_content      | Fine Root Carbon Content                 |
+#' | coarse_root_carbon_content    | Coarse Root Carbon Content               |
+#' | GWBI                          | Gross Woody Biomass Increment            |
+#' | AGB                           | Total aboveground biomass                |
+#' | time_bounds                   | history time interval endpoints          |
+
+site_ids <- design_points |>
+    pull(site_id) |>
+    unique()
+ens_ids <- 1:ensemble_size
+
+##-----TESTING SUBSET-----##
+# comment out for full run # 
+# site_ids   <- site_ids[1:5]
+# ens_ids    <- ens_ids[1:5]
+# start_year <- end_year - 1
+
+ens_dirs <- expand.grid(ens = PEcAn.utils::left.pad.zeros(ens_ids), 
+                        site_id = site_ids, 
+                        stringsAsFactors = FALSE) |>
+    mutate(dir = file.path(outdir, paste("ENS", ens, site_id, sep = "-")))
+# Check that all ens dirs exist
+existing_dirs <- file.exists(ens_dirs$dir)
+if (!all(existing_dirs)) {
+    missing_dirs <- ens_dirs[!existing_dirs]
+    PEcAn.logger::logger.warn("Missing expected ensemble directories: ", paste(missing_dirs, collapse = ", "))
+}
+
+# extract output via PEcAn.utils::read.output
+# temporarily suppress logging or else it will print a lot of file names
+logger_level <- PEcAn.logger::logger.setLevel("OFF")
+ens_results <- furrr::future_pmap_dfr(
+    ens_dirs,
+    function(ens, site_id, dir) {
+        out_df <- PEcAn.utils::read.output(
+            runid = paste(ens, site_id, sep = "-"),
+            outdir = dir,
+            start.year = start_year,
+            end.year = end_year,
+            variables = variables,
+            dataframe = TRUE,
+            verbose = FALSE
+        ) |>
+            dplyr::mutate(site_id = site_id, ensemble = as.numeric(ens)) |>
+            dplyr::rename(time = posix)
+    },
+    .options = furrr::furrr_options(seed = TRUE)
+) |>
+    group_by(ensemble, site_id, year) |>
+    #filter(year <= end_year) |> # not sure why this was necessary; should be taken care of by read.output
+    filter(time == max(time)) |> # only take last value
```

Comment:
A higher-level question to resolve in a future PR: We're reading entire years of subdaily output only to throw out all but the last timepoint. Do we have, and if not should we implement, a more efficient way of reading subsets/individual timepoints? 

Obviously harder to solve in a way general enough to handle accumulators properly ("I want the end-of-year cumsum of GPP"), but even just for state variables this seems like something I'd use a lot.